File: 20-valid-parentheses/valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
    const stack = [];
    const bracketMap = new Map([
    [')', '('],
    [']', '['], 
    ['}', '{']
]);

    for (const char of s) {
        if (bracketMap.has(char)) {
            if (stack.length && stack[stack.length - 1] === bracketMap.get(char)) {
                stack.pop();
            } else {
                return false;
            }
        } else {
            stack.push(char);
        }
    }
    return stack.length === 0;

};

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        char_map = {
            ")":"(",
            "}":"{",
            "]":"["
        }
        
        for char in s:
            if char in char_map:
                if len(stack) > 0 and stack[-1] == char_map.get(char):
                    stack.pop()
                else:
                    return False
            else:
                stack.append(char)
        
        return len(stack) == 0